Unique Pointers (Thread / Hart Rework)

rkabrick opened the pull request:
I added `@DAVE` (@donofrio) and `@Lee` (@leekillough) comments where I knew I needed you two to look specifically. However, feel free to tear the rest of it apart as you see fit.

## Minor Changes
- `RevCommon`:
  + Moved definition of `ThreadState` here
  + Changed `MAX_HARTS` to 4096 (This can be changed back but until we have someone that needs more than that there's no sense in carrying around those extra bits in the bitsets)

- `RevProc`:
  + Changed `HartToDecode` => `HartToDecodeID` to distinguish it is not an instance of `RevHart`
  + Changed `HartToExec` => `HartToExecID` to distinguish it is not an instance of `RevHart`
  + Made a `ActiveThreadID` variable so we didn't have to call into a function (ie. `GetActiveThreadID()`) which gets set at the same time `RegFile` does
  + Took ALUFault Injection Logic and put it in its own function (solely to clean up `RevProc::ClockTick`)
  + Changed `HARTS_CTE` and `HARTS_CTS` to `HartsClearToExecute` and `HartsClearToDecode`, respectively
  + Changed `GetHartID()` to `GetNextHartToDecodeID()` because `GetHartID()` sounds like you're getting the ID of the currently executing Hart
  + Added `HartHasNoDependencies(unsigned HartID)` which just returns the opposite of `AnyDependency`

- `RevFeature`:
  + Changed `HartToDecode` => `HartToDecodeID` to distinguish it is not an instance of `RevHart`
  + Changed `HartToExec` => `HartToExecID` to distinguish it is not an instance of `RevHart`

- `RevThread`:
  + Changed `GetThreadID` to just `GetID` because `Thread->GetThreadID()` seemed silly

- `RevSysCalls`:
  + Changed all the calls to `GetActiveThreadID()` to use the new `ActiveThreadID` variable

- `pthread_arg_passing test`:
  + This test was never updated to use the updated interface for `rev_pthread_create`

## Major Changes
**TL;DR**: `RevThread`'s are now exclusively used as `std::unique_ptr` and load their register state into a `RevHart`'s `RegFile` when assigned to that Hart. Below are the implications of this change

- `RevCPU`:
  + Removed global `Threads` table as there was no benefit to having each thread object live here
  + Consequently removed `AssignedThreads` because it really just added a layer of indirection and once again we gained nothing from having that global state of every thread that is on a specific RevProc. We can derive this information if need be and `RevCPU` won't ever be the one initiating a state change directly for a `RevThread`
  + When checking a Proc's `ThreadsThatChangedState` queue, it now transfers the entire queue to `RevCPU` as the `std::unique_ptr<RevThread>` needs to be moved anyways (No Thread will stay in that queue across multiple clock ticks)
  + Changed `ThreadQueue` to `ReadyThreads` which now holds the actual `RevThread` instead of just the ID. Conveyed the meaning more in my opinion
  + Changed `CheckForThreadStateChanges` => `HandleThreadStateChanges` because "checking" doesn't actually encompass what's happening

- `RevProc` / `RevHart`:
  + Removed `AssignedThreads` structure. This was effectively an unnecessary layer of indirection (ie. RevHart had the TID so every time you wanted to do anything on the thread you'd have to look it up in `AssignedThreads`)
  + `RevHart`'s now own the `std::unique_ptr<RevThread>` they are assigned and also have a `std::unique_ptr<RevRegFile>` which loads in the `RevThread's` virtual register state when assigned
  + New Bitsets introduced to convey more state
    - `HartsClearToDecode`: Formally `HART_CTS`
    - `HartsClearToExecute`: Formally `HART_CTE`
    - `IdleHarts`: A Hart is considered Idle if it does not have a `RevThread` assigned to it
    - `ValidHarts`: This is used to 'trim-the-fat' that is introduced by having `MAX_HARTS` by allowing for bitwise ops (ie. `if( IdleHarts == ValidHarts )` we know there are no threads assigned to this `RevProc`)
  + Removed `Utilization` as all we ever have to know is if there is an available (Idle) hart, or if there are no Idle harts
  + Added an `UpdateHartStatus` function which sets the values of `HartsClearToDecode` each `RevProc::ClockTick`
  + Removing a thread from a Hart is now done through `PopThreadFromHart` which returns the `std::unique_ptr<RevThread>` that was formally assigned to that Hart and marks the `RevHart` as Idle

- `RevThread`:
  + Changed `RegFile` member to `VirtRegState` and added a `using RevVirtRegState = RevRegFile;` to the class because software threads do not have a register file
    This is more of a semantic change, however, I believe it helps convey the other changes introduced in this PR
  + Changed the constructor to take in a `std::unique_ptr<RevVirtRegState>` instead of a whole bunch of values that get added to a regfile (StackPtr, ThreadPtr, etc.)

---

Overall this cuts the execution time of `BIG_LOOP` by ~50% and makes Rev a little closer aligned to how hardware operates.

rkabrick committed:
Saving Progress (Before making Hart Manager)

rkabrick committed:
Saving progress

rkabrick committed:
Saving progress. It compiles but segfaults

rkabrick committed:
Ex2 Passed

rkabrick committed:
Most tests passing

rkabrick committed:
All tests passing

rkabrick committed:
Updating HartToExec calls to HartToExecID

rkabrick committed:
Adding ThreadState to RevCommon

rkabrick committed:
Cleaning up pre-PR

rkabrick committed:
Making ActiveThreadID a variable in RevProc that gets set when RegFile does so we don't have to look it up every time

rkabrick committed:
Updating some naming inside of RevThread

rkabrick committed:
Cleaning up & Commenting RevThread.h

rkabrick committed:
Removing Pan from Proc to make rebase smoother

rkabrick committed:
Updating pthread-arg-passing test

rkabrick committed:
Fixing Merge Bugs

rkabrick committed:
Updating comments in RevCPU

rkabrick committed:
Removing HartManager file that shouldn't have ever been comitted

leekillough commented on src/RevCPU.cc:
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
   // NOTE: At this point we handle EVERY thread that changed state every cycle
-  while( !Procs[ProcID]->GetThreadsThatChangedState().empty() ){
-    auto& Thread = Procs[ProcID]->GetThreadsThatChangedState().front();
+  std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = Procs[ProcID]->TransferThreadsThatChangedState();
```
`std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = ...` might be simpler as just `auto ThreadsThatChangedState = ...`

leekillough commented on src/RevCPU.cc:
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
-    auto& Thread = Procs[ProcID]->GetThreadsThatChangedState().front();
+  std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = Procs[ProcID]->TransferThreadsThatChangedState();
+  while( !ThreadsThatChangedState.empty() ){
+    std::unique_ptr<RevThread> Thread = std::move(ThreadsThatChangedState.front());
```
See above

leekillough commented on src/RevCPU.cc:
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
 }
 
+void RevCPU::InitMainThread(uint32_t MainThreadID, const uint64_t StartAddr){
+  // @Lee: Is there a better way to get the feature info?
```
I will have to look at home

rkabrick replied:
Let me know

leekillough commented on src/RevProc.cc:
```diff
@@ -1905,36 +1885,55 @@ bool RevProc::ClockTick( SST::Cycle_t currentCycle ){
+  //        if I comment out the first block (HartToDecode based checks) the AMO tests fail
+  //        so there must be something I'm not understanding about the two. If there isn't redundancy
+  //        great, I'll uncomment this little letter I left you
+  if( RegFile->GetPC() == 0x00ull ){
```
`0` can always be compared against another type and will be automatically converted. It is never lossy. So something `0x000000000llu` (extreme example) is unnecessary and can be simply `0`.

rkabrick replied:
This is how it's been used throughout RevProc... Up to @jleidel . Not sure if there was a reason for doing it that way originally

leekillough replied:
Being the minimalist I am, I rarely compare equal or non-equal to 0 -- I would use:
```c++
if( !RegFile->GetPC() ){
```

rkabrick committed:
Adding CMakeLists to common

rkabrick committed:
Cleaning up top-level CMakeLists.txt & Removing relative paths

leekillough reviewed (commented):
I have a few other spots to review

rkabrick committed:
Moving BIG_LOOP timeout back to 90

rkabrick committed:
Removing double RevTracer.h

donofrio commented on include/RevProc.h:
```diff
@@ -190,8 +184,23 @@ class RevProc{
+  ///< RevProc: Returns true if all harts are available (ie. There is nothing executing on this Proc)
+  /// @DAVE: Can you confirm that using this function to determine if a Proc can be disabled (called in RevCPU)
+  ///        will work with the CoProc?
+  size_t HasNoBusyHarts() const { return IdleHarts == ValidHarts; }
```
I'll answer your question in a few ways... hopefully this helps.  

A thread should go to DONE state (as I understand it) once all instructions have been retired (similar to the `return` from `main`) and the `RevCoProc::IsDone() `returns true. 

A thread could certainly be Idle if it is waiting on a memory request or other dependency, this would be independent of the CoProc state.  It is also possible that the CoProc has forced a Hart to stall on Proc

A Proc could possibly be safely be disabled regardless of CoProc state **if** the CoProc is being clocked independently.  That is still a risky proposition as the CoProc can see and manipulate state within Proc and that is really not possible if the Proc is not being clocked.  The best strategy is to, as the general case, assume that the Proc cannot be disabled unless CoProc::IsDone() returns true.

rkabrick replied:
> Why `size_t` return type instead of `bool`?

Isn't that a great question 😂

rkabrick replied:
> I'll answer your question in a few ways... hopefully this helps.
> 
> A thread should go to DONE state (as I understand it) once all instructions have been retired (similar to the `return` from `main`) and the `RevCoProc::IsDone() `returns true.
> 
> A thread could certainly be Idle if it is waiting on a memory request or other dependency, this would be independent of the CoProc state. It is also possible that the CoProc has forced a Hart to stall on Proc
> 
> A Proc could possibly be safely be disabled regardless of CoProc state **if** the CoProc is being clocked independently. That is still a risky proposition as the CoProc can see and manipulate state within Proc and that is really not possible if the Proc is not being clocked. The best strategy is to, as the general case, assume that the Proc cannot be disabled unless CoProc::IsDone() returns true.

So a thread won't be removed from a Hart until all Scoreboard entries are false. When the CoProc flags a scoreboard, is it independent of the Hart's scoreboard? Also is there state beyond just the scoreboard that the CoProc can manipulate/see into?

donofrio replied:
At the moment the scoreboard and the register file are the only things the CoProc can manipulate.  It is possible, however, for a Hart to have a clear scoreboard and the CoProc to have outstanding work

rkabrick replied:
Enough said... I'll adjust the check in RevCPU

donofrio commented on include/RevProc.h:
```diff
@@ -213,17 +228,11 @@ class RevProc{
 
   // If a given assigned thread experiences a change of state, it sets the corresponding bit
-  std::queue<std::shared_ptr<RevThread>> ThreadsThatChangedState; ///< RevProc: used to signal to RevCPU that the thread assigned to HART has changed state
+  std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState; ///< RevProc: used to signal to RevCPU that the thread assigned to HART has changed state
```
Kept the same name, eh? ;)

rkabrick replied:
Changing to Homer

donofrio commented on include/RevThread.h:
```diff
@@ -76,58 +46,100 @@ class RevThread {
+  // Arguments could be made to keep it here or put it there
+  void SetMarkLoadComplete(std::function<void(const MemReq&)> func){
+    VirtRegState->SetMarkLoadComplete(func);
+  }
```
A LSQ manages requests for multiple threads/harts but is tied to a single Proc, and as a Hart is also tied to a Proc it may make more sense to put it in RevHart, especially if threads can migrate between Procs - having LSQ in RevHart would remove the chance of a thread ever referencing a LSQ outside the Proc it is are running on.

rkabrick replied:
Moved... the LSQueue & MarkLoadComplete are now parts of the RevHart constructor. Everything passes but please review the new implementation in RevHart

donofrio commented on src/RevProc.cc:
```diff
@@ -1905,36 +1885,55 @@ bool RevProc::ClockTick( SST::Cycle_t currentCycle ){
+  // @DAVE: I feel like the below two blocks of code have some redundancy but its working and
+  //        if I comment out the first block (HartToDecode based checks) the AMO tests fail
+  //        so there must be something I'm not understanding about the two. If there isn't redundancy
+  //        great, I'll uncomment this little letter I left you
```
Dear Ryan,

Thank you for your letter. There is some redundancy - in order to be done we need to check that there is nothing stalled waiting to decode AND that even if there is nothing to decode there is no thread with an instruction currently in flight causing us to exit early. I am curious, what if you add ` HartHasNoDependencies(HartToDecodeID)` as an additional condition in the 2nd `if` does that work?

` if( HartToExecID != _REV_INVALID_HART_ID_ && !IdleHarts[HartToExecID]
        && HartHasNoDependencies(HartToExecID)
        &&  HartHasNoDependencies(HartToDecodeID)
        && ( (nullptr == coProc) || (coProc && coProc->IsDone() )
        ) ){
      std::unique_ptr<RevThread> ActiveThread = PopThreadFromHart(HartToDecodeID);
      ActiveThread->SetState(ThreadState::DONE);
      HartsClearToExecute[HartToExecID] = false;
      HartsClearToDecode[HartToExecID] = false;
      IdleHarts[HartToExecID] = true;
      ThreadsThatChangedState.emplace(std::move(ActiveThread));
    }`

rkabrick replied:
You're a legend! Running tests now and it's looking like that works

rkabrick replied:
nvm those tests time out

donofrio commented on src/RevCPU.cc:
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
+    }
+    // Remove the pointer to the thread as it was handled
+    ThreadsThatChangedState.pop();
   }
```
At some point we may need to create Current vs. Next state for threads, but I think given the separate ClockTick functions in RevProc and RevCPU, we should be OK as long as we are careful in how the updates are done

rkabrick replied:
That seems like something that is definitely possible to achieve with the current implementation (ie. we haven't coded ourselves into a corner) so if it's cool with you I'll leave it as is in this PR and down the road we can revisit

rkabrick committed:
Minor changes from PR feedback

rkabrick commented on src/RevCPU.cc:
```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
-  }
-  output.verbose(CALL_INFO, 4, 0, "Initializing Thread %" PRIu32 "\n", TID);
+  RegState->SetX(RevReg::gp, gp);
+  RegState->SetX(RevReg::s0, gp); // This can be used as the fp reg per ABI
```
In your `enum class` is it possible to have an `fp` register point to the same value as the `s0` register? Or is it a single key only

leekillough replied:
> In your `enum class` is it possible to have an `fp` register point to the same value as the `s0` register? Or is it a single key only

Already `fp` registers and `x` (gpr) registers have tags which are the same values. What `fp` or `gpr` register do you want added/changed?

The names of the registers are their ABI names explicitly specified in the RISC-V Architecture SPEC (like `a0`, `sp`, `gp`, `ra`). This is intentional because those names are more familiar. Is there a name you need which is not a part of the ABI naming convention?

You can use integers instead of `RevReg` enums to give a raw register number, which is how `rs1`, `rs2`, `rs3` and `rd` are handled. The functions taking registers are templated to accept either integers or `RevReg` and are explicitly cast to `size_t`  before using them as a raw register number.

leekillough replied:
> In your `enum class` is it possible to have an `fp` register point to the same value as the `s0` register? Or is it a single key only

Sorry I misunderstood. `fp` is the official ABI name of frame pointer, which is why it's used in the `RevReg` enum.

I was referring to your **comment** after `//`. "Frame pointer" should be used in the comment instead of `fp` which is ambiguous and means "floating point" much more often.

leekillough replied:
"This can be used as the fp reg..." should be "This can be used as the frame pointer reg"

leekillough replied:
"fp reg" 99% of the time means "floating point register", so "frame pointer" should be worded explicitly to avoid confusion.

`RevReg::fp` is the ABI's official name for the frame pointer register. The `RevReg::` prefix makes that unambiguous. Bare "fp" used in comments almost always means floating-point.

leekillough replied:
... and since there are "fp registers", "the fp register" is even more ambiguous.

rkabrick replied:
Ah yeah those are all very valid points

rkabrick commented on src/RevCPU.cc:
```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
-
+// - Adds it's ThreadID to the ReadyThreads to be scheduled
+void RevCPU::InitThread(std::unique_ptr<RevThread> ThreadToInit){
+  // @Lee, can this be a reference? I'm not looking to transfer ownership of the regfile pointer
```
Alright I changed it to &&. Interesting discussion though. Thank you for the insight!

leekillough replied:
In the old code you were constructing a new `unique_ptr` for the function argument, and because `unique_ptr` has no copy constructor, only a move constructor, you had to call the function with `std::move(arg)`. Once in the function, the argument/parameter local variable has ownership of the `unique_ptr`. When you transfer it to its new destination, you have to use `std::move` again.

With `&&` rvalue reference for the function parameter, you are requiring that it only bind to rvalues, so lvalues have to be converted to rvalues by using `std::move` at the call site. Within the function, the parameter name acts as an lvalue to the referent. You have to use `std::move` again to transfer it to its new destination.

The difference between old and new is that a local temporary `unique_ptr` and moves in and out of it were avoided -- now you just use a rvalue reference for the parameter. It still only accepts rvalue arguments, and it still must be converted to a rvalue when it is transferred to somewhere else.

rkabrick commented on src/RevCPU.cc:
```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
+// This should not be called without first checking if the Proc has an IdleHart
+void RevCPU::AssignThread(std::unique_ptr<RevThread> ThreadToAssign, unsigned ProcID){
+  output.verbose(CALL_INFO, 4, 0, "Assigning Thread %" PRIu32 " to Processor %" PRIu32 "\n", ThreadToAssign->GetID(), ProcID);
+  Procs[ProcID]->AssignThread(std::move(ThreadToAssign));
```
In this case, you are transferring ownership of the `ThreadToAssign` to a `RevProc`. So it will not be used again until the `std::unique_ptr<RevThread>` is returned via the `ThreadsThatChangedState` handling. Does this affect your comment?

leekillough replied:
I don't know what comment "above" I was referring to, but I think it involves the rvalue reference and move stuff, so you should change the function to:
```c++
void RevCPU::AssignThread(std::unique_ptr<RevThread>&& ThreadToAssign, unsigned ProcID){
```
The rest of the function stays the same. You are avoiding creating a temporary local `ThreadToAssign` and are using a rvalue reference to the replacement directly.

rkabrick committed:
PR Feedback

rkabrick committed:
Lower default verbosity on memh-pthreads test

rkabrick committed:
Merge branch 'relative-header-fix' into unique-ptrs

rkabrick committed:
PR Fixes & Merging relative-header-fix

leekillough commented on include/RevThread.h:
```diff
@@ -76,58 +46,100 @@ class RevThread {
+  // @DAVE: Potentially move these to RevHart and have them set it when it's assigned?
+  // Arguments could be made to keep it here or put it there
+  void SetLSQueue(std::shared_ptr<std::unordered_map<uint64_t, MemReq>> lsq){
+    VirtRegState->SetLSQueue(lsq);
```
If you a reference, you'll avoid an atomic increment/decrement of the reference count which occurs when creating and destroying the `lsq` local parameter.
```c++
  void SetLSQueue(const std::shared_ptr<std::unordered_map<uint64_t, MemReq>>& lsq){
```

leekillough commented on include/RevThread.h:
```diff
@@ -76,58 +46,100 @@ class RevThread {
+  // @DAVE: Potentially move these to RevHart and have them set it when it's assigned?
+  // Arguments could be made to keep it here or put it there
+  void SetMarkLoadComplete(std::function<void(const MemReq&)> func){
+    VirtRegState->SetMarkLoadComplete(func);
```
Copying a `std::function` requires copying all of its captured variables, allocating memory, etc. Moving it should be faster.

rkabrick committed:
PR Feedback... Moving LSQueue setting to RevHart instead of RevThread

rkabrick committed:
More PR Feedback fixups

leekillough commented on src/RevCPU.cc:
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
 
+void RevCPU::InitMainThread(uint32_t MainThreadID, const uint64_t StartAddr){
+  // @Lee: Is there a better way to get the feature info?
+  std::unique_ptr<RevRegFile> MainThreadRegState = std::make_unique<RevRegFile>(Procs[0]->GetRevFeature());
```
`Procs[0]` seems to be arbitrary. Is there a way to figure out which `Procs` to use here?

rkabrick replied:
It is semi-arbitrary. I used 0 because we ALWAYS know that there will be at least one Proc (Core ;)). Further, we also know that the main thread will always end up on the first core. 

I still don't like it... but currently the RevFeature is created inside of RevProc so that's the only way I can think to gather the info.

donofrio replied:
Right now we assume a homogeneous CPU, so this is OK.  To get to a heterogeneous CPU (like a big.LITTLE) we'll need to do some rearchitecting and the CPU will need to have some description of what features (RV and other) are in which Procs. I don't think we need to solve that here, or hold up your PR for that reason

leekillough commented on test/threading/pthreads/pthread-arg-passing/pthread-arg-passing.c:
```diff
@@ -1,13 +1,19 @@
-#define assert(x) if (!(x)) { asm(".byte 0x00"); asm(".byte 0x00"); asm(".byte 0x00"); asm(".byte 0x00"); }
+#define assert(x)                                                              \
+  if (!(x)) {                                                                  \
+    asm(".byte 0x00");                                                         \
```
`asm(".word 0");` instead of four `asm('.byte 0x00");` would be simpler

rkabrick committed:
Adding missing comments

leekillough committed:
Simplify code; use std::vector instead of std::queue because it's faster

leekillough reviewed (changes requested):
See https://github.com/tactcomplabs/rev/pull/185

rkabrick committed:
Merge pull request #185 from tactcomplabs/Lee-unique-ptrs

Simplify code; use std::vector instead of std::queue because it's faster

rkabrick committed:
Making Pipeline a deque because the erasing of the first element ends up being an O(n) operation to shift everything down

rkabrick committed:
moving ALUFault to its own function

rkabrick committed:
Inline experiments

rkabrick committed:
Revert "Inline experiments"

This reverts commit 9f4854abf6776f8408613cf477242d437bc0a655.

rkabrick committed:
Simplifying DependencySet and moving to header so inlined

rkabrick committed:
Increase timeout of BIG_LOOP by 10 Seconds

Original (Pre-RevHart) timeout was 90... post RevHart we were seeing executions take 2 minutes. The average time now with this PR is ~91 seconds.

rkabrick committed:
Merge branch 'devel' into unique-ptrs

rkabrick committed:
fixing merge bug
